packages/grid/_modules_/grid/components/panel/filterPanel/GridFilterForm.tsx

```diff
@@ -87,27 +87,89 @@ function GridFilterForm(props: GridFilterFormProps) {
     (event: SelectChangeEvent) => {
       const columnField = event.target.value as string;
       const column = apiRef.current.getColumn(columnField)!;
-      const newOperator = column.filterOperators![0];
 
-      applyFilterChanges({
-        ...item,
-        value: undefined,
-        columnField,
-        operatorValue: newOperator.value,
-      });
+      // try to keep the same operator when column change
+      const newOperator =
+        column.filterOperators!.find((operator) => operator.value === item.operatorValue) ||
+        column.filterOperators![0];
+
+      if (column.type === 'singleSelect') {
```

Comment:
We shouldn't have code specific to one column type here. Why it was added?

Reply:
For two reasons:
- Its InputComponent is the same as the InputComponent for string or number. Then the comparison of InputComponent will not detect if we switch from a "string" to a "singleSelect"
- It has an additional parameter: `valueOptions` so I need a specific case to check if the value is still available in the new column

Reply:
> Its InputComponent is the same as the InputComponent for string or number.

One more reason to extract into another component the logic for `singleSelect`. https://github.com/mui-org/material-ui-x/pull/3018#discussion_r739439422

> It has an additional parameter: valueOptions so I need a specific case to check if the value is still available in the new column

`GridFilterInputValue` should sanitize the value it receives. This includes, if it's a `singleSelect`, checking if the value is included in the options.

Reply:
Ok, splitting components makes sense. I'm opening another PR for that